recipes/lora_dpo_single_device.py

```diff
@@ -0,0 +1,512 @@
+        ).squeeze(2)
+
+        return (per_token_log_probs * loss_mask).sum(-1)
+
+    def train(self) -> None:
+        """
+        The core training loop.
+        """
+
+        # self.epochs_run should be non-zero when we're resuming from a checkpoint
+        for curr_epoch in range(self.epochs_run, self.total_epochs):
+            # Update the sampler to ensure data is correctly shuffled across epochs
+            # in case shuffle is True
+            self._sampler.set_epoch(curr_epoch)
+            for idx, batch in enumerate(pbar := tqdm(self._dataloader)):
+                if (
+                    self.max_steps_per_epoch is not None
+                    and (idx // self._gradient_accumulation_steps)
+                    == self.max_steps_per_epoch
+                ):
+                    break
+
+                (
+                    policy_chosen_log_probs,
+                    policy_rejected_log_probs,
+                    policy_chosen_logits,
+                    policy_rejected_logits,
+                ) = self.concatenated_forward(self._model, batch)
+
+                with torch.no_grad(), disable_adapter(self._model):
+                    (
+                        reference_chosen_log_probs,
+                        reference_rejected_log_probs,
+                        _,
+                        _,
+                    ) = self.concatenated_forward(self._model, batch)
+
+                loss, chosen_rewards, rejected_rewards = self._loss_fn(
+                    policy_chosen_log_probs,
+                    policy_rejected_log_probs,
+                    reference_chosen_log_probs,
+                    reference_rejected_log_probs,
+                )
+                loss = loss.mean()
+                reward_accuracies = (chosen_rewards > rejected_rewards).float()
+                if self.total_training_steps % self._log_every_n_steps == 0:
+                    pbar.set_description(f"{curr_epoch+1}|{idx+1}|Loss: {loss.item()}")
+                    self._metric_logger.log_dict(
+                        {
+                            "loss": loss.item(),
+                            "lr": self._optimizer.param_groups[0]["lr"],
+                            "rewards/chosen": chosen_rewards.mean().cpu(),
+                            "rewards/rejected": rejected_rewards.mean().cpu(),
+                            "rewards/accuracies": reward_accuracies.mean().cpu(),
+                            "rewards/margins": (chosen_rewards - rejected_rewards).mean().cpu(),
+                            "log_probs/rejected": policy_rejected_log_probs.detach().mean().cpu(),
+                            "log_probs/chosen": policy_chosen_log_probs.detach().mean().cpu(),
+                            "logits/rejected": policy_rejected_logits.detach().mean().cpu(),
+                            "logits/chosen": policy_chosen_logits.detach().mean().cpu(),
+                            "gpu_resources": torch.cuda.memory_allocated(),
```

Comment:
I'm curious, do we need all of these in general? E.g. mean of logits feels like it may be covered better by log_probs anyways. (Btw I think we should think about a better way to make our metric logging more configurable in general, so it's not necessarily specific to your DPO recipe)

Reply:
> I'm curious, do we need all of these in general? E.g. mean of logits feels like it may be covered better by log_probs anyways.

I have the same doubts, too, you can check my [training log](https://github.com/pytorch/torchtune/pull/645#issuecomment-2041861215) to see the difference. (by the way I'm still confused)